src/domain/orders/details/returns/index.tsx

```diff
@@ -0,0 +1,295 @@
+import RMASelectProductTable from "../../../../components/organisms/rma-select-product-table"
+import LayeredModal, {
+  LayeredModalContext,
+} from "../../../../components/molecules/modal/layered-modal"
+
+const removeNullish = (obj) =>
+  Object.entries(obj).reduce((a, [k, v]) => (v ? ((a[k] = v), a) : a), {})
+
+const ReturnMenu = ({ order, onReturn, onDismiss, toaster }) => {
+  const layoutmodalcontext = useContext(LayeredModalContext)
+
+  const [submitting, setSubmitting] = useState(false)
+  const [refundEdited, setRefundEdited] = useState(false)
+  const [refundable, setRefundable] = useState(0)
+  const [refundAmount, setRefundAmount] = useState(0)
+  const [toReturn, setToReturn] = useState({})
+  const [quantities, setQuantities] = useState({})
+  const [useCustomShippingPrice, setUseCustomShippingPrice] = useState(false)
+
+  const [shippingLoading, setShippingLoading] = useState(true)
+  const [shippingOptions, setShippingOptions] = useState([])
+  const [noNotification, setNoNotification] = useState(order.no_notification)
+  const [shippingPrice, setShippingPrice] = useState<number>()
+  const [shippingMethod, setShippingMethod] = useState(null)
+
+  const [allItems, setAllItems] = useState<any[]>([])
+
+  useEffect(() => {
+    if (order) {
+      setAllItems(filterItems(order, false))
+    }
+  }, [order])
+
+  useEffect(() => {
+    Medusa.shippingOptions
+      .list({
+        region_id: order.region_id,
+        is_return: true,
+      })
+      .then(({ data }) => {
+        setShippingOptions(data.shipping_options)
+        setShippingLoading(false)
+      })
+  }, [])
+
+  useEffect(() => {
+    const items = Object.keys(toReturn).map((t) =>
+      allItems.find((i) => i.id === t)
+    )
+    const total =
+      items.reduce((acc, next) => {
+        return (
+          acc +
+          (next.refundable / (next.quantity - next.returned_quantity)) *
+            toReturn[next.id].quantity
+        )
+      }, 0) - (shippingPrice || 0)
+
+    setRefundable(total)
+
+    setRefundAmount(total)
+  }, [toReturn, quantities, shippingPrice])
+
+  const onSubmit = () => {
+    const items = Object.entries(toReturn).map(([key, value]) => {
+      const clean = removeNullish(value)
+      return {
+        item_id: key,
+        ...clean,
+      }
+    })
+
+    const data = {
+      items,
+      refund: Math.round(refundAmount),
+      no_notification:
+        noNotification !== order.no_notification ? noNotification : undefined,
+    }
+
+    if (shippingMethod) {
+      data.return_shipping = {
+        option_id: shippingMethod.value,
+        price: shippingPrice / (1 + order.tax_rate / 100),
+      }
+    }
+
+    if (onReturn) {
+      setSubmitting(true)
+      return onReturn(data)
+        .then(() => onDismiss())
+        .then(() => toaster("Successfully returned order", "success"))
+        .catch((error) => toaster(getErrorMessage(error), "error"))
+        .finally(() => setSubmitting(false))
```

Comment:
We should consider moving the promise handling to the parent component as part of the `onSuccess` and `onError` functions of the new admin hooks. Let me know what you think?

If we choose to do this, there are a couple of more places to change it.

Reply:
hmm, it's passed directly from `useMedusa('orders')`, I dont know if we should change it with this pr or fix all of the methods at once?

Reply:
Since we have the order in this component already, would it be possible for us to use the hook in this component instead?
```
const returnOrder = useAdminRequestReturn(order.id)

return.mutate(.....
```

Reply:
@pKorsholm

Reply:
this does work and i have pushed it, though it doens't update the view now

Reply:
@olivermrbl